lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <algorithm>
+#include <map>
+
+#include "ShardTracker.h"
+#define MAX_SHARD_SIZE 100
+
+namespace magma {
+
+ShardTracker::ShardTracker() {
+  number_of_shards = 0;
+}
+
+int ShardTracker::add_ue() {
+  int shard_id;
+  if (shards.size() == 0) {
+    shard_id = 0;
+    shards.push_back(1);
+    number_of_shards++;
+  } else {
+    for (auto i = 0; i < number_of_shards; i++) {
+      if (shards[i] < MAX_SHARD_SIZE) {
+        shards[i]++;
+        return i;
+      }
+    }
+    shard_id = number_of_shards++;
+    shards.push_back(1);
+  }
+  return shard_id;
+}
+
+void ShardTracker::remove_ue(int shard_id) {
+  shards[shard_id]--;
```

Comment:
I might be missing something but how are we guaranteeing that the shard_id index is available?

Reply:
So the shard_id index is available based on the number of UEs it has. I'll add a comment for this, but every time the UE is added a check is made for the number of UEs at a particular index before addition.